### DISCUSSION TITLE
Bad verify token

### DISCUSSION THREAD

**Discussion by t0vers** [fastapi-users/fastapi-users#1398] [Q&A] [answered]
Hi! when registering, I send the verification token to an email. after that, I manually copy it and send it to the verify endpoint. however, it always returns me 400 VERIFY_USER_BAD_TOKEN

```
class UserManager(BaseUserManager[UserCreate, UserRead]):
    verification_token_secret = "SECRET"
    verification_token_audience = "fastapi-users:verify"
    verification_token_lifetime_seconds = 3600

    async def on_after_register(self, user: UserRead, request: Request):
        async with AsyncClient() as client:
            response = await client.post(
                f"{request.base_url}auth/request-verify-token",
                json={"email": user.email}
            )
            response.raise_for_status()

    async def parse_id(self, id_str: str) -> Optional[str]:
        return id_str

    async def send_verify_message(self, email: str, token: str):
        message = MessageSchema(
            subject="Verify your email",
            recipients=[email],
            body=f"{token}",
            subtype="html"
        )
        await fast_mail.send_message(message)

    async def on_after_request_verify(self, user: UserRead, token: str, request: Request):
        await self.send_verify_message(user.email, token)
        print(f"Verification requested for user {user.id}. Verification token: {token}")
```

**Comment by hgalytoby** [comment]
I didn't have the problem you mentioned when I just tested it.

Have you tried copying the token from the log first instead of the token in the email?

**Reply by t0vers** [reply]
I tried both options, it didn`t help

**Reply by hgalytoby** [reply]
> I tried both options, it didn`t help

I think you have made a mistake.


https://github.com/fastapi-users/fastapi-users/assets/50397689/ccfd3564-109f-42bb-b684-e828c1cba4a5

**Reply by t0vers** [reply]
> > I tried both options, it didn`t help
> 
> I think you have made a mistake.
> 
>  20240616_031613.mp4


https://github.com/fastapi-users/fastapi-users/assets/171922396/7526fa9f-f67d-454f-bbd3-886b938de4f5

**Reply by hgalytoby** [reply, marked as answer]
It looks fine from your video.

This is my guess.

I see in your video that your docker project is hotels and I see this project in your repo.

I see that your user model is Integer

https://github.com/t0vers/hotels/blob/master/user-service/database/database.py#L22

But I see from your code that you are using

```py
async def parse_id(self, id_str: str) -> Optional[str]:
    return id_str
```

From the verify code, I think you may have a problem with parse_id.

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py#L341

You should use IntegerIDMixin

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py#L695

```py
class UserManager(IntegerIDMixin, BaseUserManager[UserCreate, UserRead]):
```

I try to try the same scenario as you myself, using the User model using Integer with the `parse_id` you wrote, and I do have problems with the validation failing.

Using `parse_id` in the verify method is a synchronous method, but the `parse_id` you wrote is a asynchronous method, and even if you change it to a synchronous method, it will still fail because str and int are not equal.

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/manager.py#L346

But if it's not a problem with parse_id, could you please print out all the judgments in the verify method, so you should be able to see where the validation fails.

**Reply by t0vers** [reply]
Yes, it helped me solve the problem. Thank you very much, you helped me a lot!
Btw, I saw that you cannot log in without verification. can you tell me how this functionality was implemented?

**Reply by hgalytoby** [reply]
https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/router/auth.py#L17

https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/router/auth.py#L64

https://fastapi-users.github.io/fastapi-users/latest/configuration/routers/auth/#optional-user-verification
